### DISCUSSION TITLE
Using IServiceProvider in C# with async

### DISCUSSION THREAD

**Discussion by karpikpl** [pulumi/pulumi#10623] [Pulumi in Practice] [answered]
I'm little confused by the `Pulumi.Deployment` class.

It seems to support dependency injection by allowing some methods take `IServiceProvider` added in https://github.com/pulumi/pulumi/pull/5723 by @nathanwinder-perficient but at the same time there seems to be a recommendation to use `async/await`, which is impossible when using `Stack`, because program flow uses `Stack` constructor as an entry point.

There was a another question about it asked In April - https://github.com/pulumi/pulumi/discussions/9495#discussioncomment-3544903

So looking at `Deployment` class we can either:
1. have dependency injection but no async/await
2. have async/await but no dependency injection
?
@mikhailshilkov 

What I'm missing is one good pulumi example where:
1. Async/Await is used
2. Everything is not created in a single constructor
3. Unit tests are present
4. Pulumi is using dependency injection 

Maybe a solution for tests/injection would be creating:
```csharp
public abstract AsyncStack
{
     public abstract Task CreateAsync();
}
```

with `Deployment` having a method to run it?
```csharp

            Task<int> IRunner.RunAsync<TStack>(IServiceProvider serviceProvider)
            {
                IServiceProvider serviceProvider2 = serviceProvider;
                if (serviceProvider2 == null)
                {
                    throw new ArgumentNullException("serviceProvider");
                }
                return RunAsync(() => (serviceProvider2.GetService(typeof(TStack)) as TStack) ?? throw new ApplicationException(string.Format("Failed to resolve instance of type {0} from service provider. Register the type with the service provider before calling {1}.", typeof(TStack), "RunAsync")));
            }

            Task<int> IRunner.RunAsync<TStack>()
            {
                return RunAsync(() => new TStack().CreateAsync());
            }
```

**Comment by mikhailshilkov** [comment]
Copying our .NET folks @Zaid-Ajaj and @Frassle
(how do I do this without posting an answer? I'm a bit confused about Discussions)

**Comment by Zaid-Ajaj** [comment, marked as answer]
Hi @karpikpl, first of all I want to point out that the use of dependency injection with Pulumi is possible with or without `IServiceProvider`. It is only needed when you really complex object graphs and you don't want to bother with instantiating them by hand. 

I would personally argue that you actually don't need `IServiceProvider` and would be better off using dependency injection in its simplest form: passing parameters. In any case, the following example can be used with or without an `IServiceProvider` as I will show you. Here is our use case:

You want to create a Pulumi program which populates an AWS bucket with bucket objects from [The Cats API](https://thecatapi.com/). Now we also want to unit test this program to assert that it is indeed populating a single bucket object per returned entry from the cats API. This means we want different implementation of an interface that retrieves images of cats, one is the "real" implementation that goes out to the API and another that is used for our tests returning mock entries. 

Here is our data model:

```cs
public class CatInfo
{
    public string Id { get; set; }
    public string ImageUrl { get; set; }
}

public interface ICatApi
{
    Task<List<CatInfo>> FetchCats(int limit);
}
```
The "real" implementation, it looks like this:
```cs
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Net.Http;
using Newtonsoft.Json.Linq;

public class RealCatsApi : ICatApi
{
    public async Task<List<CatInfo>> FetchCats(int limit)
    {
        var client = new HttpClient();
        var catsJson = await client.GetStringAsync($"https://api.thecatapi.com/v1/images/search?limit={limit}");
        var cats = new List<CatInfo>();
        foreach (var catJson in JArray.Parse(catsJson))
        {
            cats.Add(new CatInfo
            {
                Id = catJson["id"]?.ToObject<string>()!,
                ImageUrl = catJson["url"]?.ToObject<string>()!
            });
        }

        return cats;
    }
}
```
Now our implementation of the Pulumi program will not depend on the `RealCatsApi` but instead it depends on `ICatsApi` as 
follows:

> NOTE: depending on `ICatsApi` means taking `ICatsApi` as parameter

```cs
async Task<Dictionary<string, object?>> DeloyCatsBucket(ICatApi catsApi)
{
     var cats = await catsApi.FetchCats(10);
     var bucket = new Bucket("cats");
     foreach (var cat in cats)
     {
         new BucketObject($"cat-{cat.Id}", new()
         {
            Bucket = bucket.BucketName,
            Source = new RemoteAsset(cat.ImageUrl)
         });
     }

    return new Dictionary<string, object?>
    {
       ["bucketName"] = bucket.Id
    };
}

```

The interesting part it the entry point which serves as the _composition root_ of the deployment:

### Using _simple_ dependency injection without IServiceProvider:
```cs
return await Deployment.RunAsync(async () =>
{
    // instantiate a real implementation
    var catsApi = new RealCatsApi();
    // "inject" catsApi into the DeloyCatsBucket function (our Pulumi program)
    return await DeloyCatsBucket(catsApi);
});
```
Now `DeloyCatsBucket` is unit testable as it, you only need to put it in a class like this:
```cs
public class Infra
{
   public static async Task<Dictionary<string, object?>> DeloyCatsBucket(ICatApi catsApi)
   {
         // implementation
   }
}

return await Deployment.RunAsync(async () =>
{
    // instantiate a real implementation
    var catsApi = new RealCatsApi();
    // "inject" catsApi into the DeloyCatsBucket function (our Pulumi program)
    return await Infra.DeloyCatsBucket(catsApi);
});
```
Then from your unit tests
```cs
var (resources, outputs) = await Deployment.TestAsync(mocks, options, async () =>
{
    // instantiate a mock API
    var fakeCatsApi = new Mock<ICatsApi>();
    // setup the mock to return empty result
    fakeCatsApi.Setup(catsApi => catsApi.FetchCats(It.Any<int>()))
                        .Returns(Task.FromResult(new List<CatInfo> { }));
   
   return await Infra.DeloyCatsBucket(fakeCatsApi.Object);
});

Assert.Empty(resources.OfType<BucketObject>(), "No bucket object was created");
```
### Using `IServiceProvider`:
Here you need first an implementation of the service provider but it can be whatever `IServiceProvider` you have at hand. I am using a bare-minimum one here but I can imagine more advanced ones with better registration functionality
```cs
class CatsServiceProvider : IServiceProvider
{
    public object GetService(Type serviceType)
    {
        if (serviceType == typeof(ICatsApi))
        {
            return new RealCatsApi();
        }

        return null;
    }
}

return await Deployment.RunAsync(async () =>
{
    var serviceProvider = new CatsServiceProvider();
    // require an implementation of ICatsApi
    var catsApi = serviceProvider.GetService(typeof(ICatsApi));
    // "inject" catsApi into the DeloyCatsBucket function (our Pulumi program)
    return await DeloyCatsBucket(catsApi);
});
```
Same idea for unit testing when using `IServiceProvider`, you would register mock implementations of your dependencies into a service provider, then require those in tests
```cs
var (resources, outputs) = return await Deployment.TestAsync(mocks, options, async () =>
{
    var serviceProvider = new TestCatsServiceProvider();
    // require an implementation of ICatsApi
    var catsApi = serviceProvider.GetService(typeof(ICatsApi));
    // "inject" catsApi into the DeloyCatsBucket function (our Pulumi program)
    return await DeloyCatsBucket(catsApi);
});
```
### "Scaling" to more dependencies 

The ideas here are the same if you had even more dependencies: add more parameters and keep it simple. Dependency injection doesn't have to be a mess 😄 
```cs
async Task<Dictionary<string, object?>> DeloyCatsBucket(ICatApi catsApi, IDogsApi dogsApi, IFileSystem fs)
{
    // use the dependencies here
}
```
if your function has "too many" dependencies, you can always break it down into more functions.

Hope my explanation was clear! Let me know if you questions about it.

**Reply by karpikpl** [reply]
Hey @Zaid-Ajaj 
this makes perfect sense, I think it got confused by the `IServiceProvider` deployment method. I assumed I have to use it, but as you pointed out, I should be testing the `Stack` not the Deployment, so I can control my dependencies there.

Since I have you here :)
You're using an async method `public static async Task<Dictionary<string, object?>> DeloyCatsBucket(ICatApi catsApi)`, which also makes sense for me, but does it mean the **recommended** approach is NOT to inherit from `Stack` anymore?

Most templates and examples deploy everything in one anonymous method. Some examples I've seen outside of pulumi repos, inherit from `Stack` and deploy using `Deploy<TStack>` methods.

I'm trying to decide between having everything in constructors:
```csharp
public class MyStack: Stack
{
      public MyStack()
     {
         var appInsights = new Components.Diagnostics("insights", new Components.Diagnostics.DiagnosticsArgs
        {
            ResourceGroupName = resourceGroup.Name,
            Tags = tags
        });

        var sql = new Components.Sql("sql", new Components.Sql.SqlArgs
        {
            ResourceGroupName = resourceGroup.Name,
            SqlAdAdminLogin = config.SqlAdAdminLogin!,
            SqlAdAdminPassword = config.SqlAdAdminPassword!,
            IsDBSchemaDeploymentEnabled = config.IsDBSchemaDeploymentEnabled,
            Tags = tags
        });
     }
}
```

and having my stack created in an async method, with all resources created in `ExecuteAsync` methods:
```csharp
public class CoreExStack
{
    public static async Task<IDictionary<string, object?>> ExecuteStackAsync()
   {
         var config = await StackConfiguration.CreateConfiguration();
        Log.Info("Configuration completed");

        var tags = new InputMap<string> { { "App", "CoreEx" } };

        // Create an Azure Resource Group
        var resourceGroup = new ResourceGroup($"coreEx-{Pulumi.Deployment.Instance.StackName}", new ResourceGroupArgs
        {
            Tags = tags
        });

        var storage = new Components.Storage("sa", new Components.Storage.StorageArgs
        {
            ResourceGroupName = resourceGroup.Name,
            Tags = tags
        });
        await storage.ExecuteAsync();
       }
}
```

**Reply by Zaid-Ajaj** [reply]
> does it mean the recommended approach is NOT to inherit from Stack anymore?

It is simply easier to not use `Stack`. Using functions gives you immediate benefits:
 - Straightforward to async/await because no constructor limitations
 - No need to define output properties before hand and mark them with [Output]
 - Simpler to use with top-level statements in C# where you can start writing resources inside `Deployment.RunAsync`

We are not going to deprecate `Stack` any time soon but newer examples and templates are moving towards the functional approach.

> Some examples I've seen outside of pulumi repos, inherit from Stack and deploy using Deploy<TStack> methods.

I think that is because mostly we haven't been advocating for the functional approach before and everyone who followed our docs used `Stack`. Now that is not a bad thing, but IMO most of the these examples can be greatly simplified and reduced down to a function call 😄 

> I'm trying to decide between having everything in constructors:

It depends on how complicated your constructors can be. Do you have something to `await` like a function invoke? then a function is the way to go. Also they don't always have to return `Task<Dictionary<string, object?>>`, if you don't have outputs to return, have the return type be `Task` and that should do it.
